Issue: Priority queue, delay queue and timing-wheel with data race issue

* go 版本 1.22.0
* 系统版本 windows 11

```
设备名称	kvii
处理器	Intel(R) Core(TM) i5-9300H CPU @ 2.40GHz   2.40 GHz
机带 RAM	32.0 GB
设备 ID	3A3E54E1-1AF3-41C4-9E01-93A303ED3F64
产品 ID	00330-80000-00000-AA250
系统类型	64 位操作系统, 基于 x64 的处理器
笔和触控	没有可用于此显示器的笔或触控输入

版本	Windows 11 专业版
版本	23H2
安装日期	‎2022/‎5/‎14
操作系统版本	22635.3209
体验	Windows Feature Experience Pack 1000.22687.1000.0
```

日志输出

```sh
$ go test -race .\lib\queue\

2024/02/23 16:13:58 maxprocs: Leaving GOMAXPROCS=8: CPU quota undefined
==================
WARNING: DATA RACE
Read at 0x00c00018c038 by goroutine 9:
  github.com/benz9527/xboot/lib/queue.(*ArrayPriorityQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).Len()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:105 +0x4c
  github.com/benz9527/xboot/lib/queue.(*ArrayPriorityQueue[*github.com/benz9527/xboot/lib/queue.employee]).Len()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:104 +0x17
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).popIfExpired()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:45 +0x43
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).poll()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:92 +0x15a
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_windows.go:21 +0x144
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[*github.com/benz9527/xboot/lib/queue.employee]).PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_windows.go:11 +0x54
  github.com/benz9527/xboot/lib/queue.TestArrayDelayQueue_PollToChan.gowrap1()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_test.go:38 +0x6f

Previous write at 0x00c00018c038 by goroutine 8:
  github.com/benz9527/xboot/lib/queue.(*arrayPQ[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).Push()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:97 +0x1a7
  github.com/benz9527/xboot/lib/queue.(*arrayPQ[*github.com/benz9527/xboot/lib/queue.employee]).Push()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:88 +0x4a
  container/heap.Push()
      C:/Program Files/Go/src/container/heap/heap.go:52 +0x45
  github.com/benz9527/xboot/lib/queue.(*ArrayPriorityQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).Push()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:117 +0xa4
  github.com/benz9527/xboot/lib/queue.(*ArrayPriorityQueue[*github.com/benz9527/xboot/lib/queue.employee]).Push()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:116 +0x2e
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).Offer()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:65 +0x206
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[*github.com/benz9527/xboot/lib/queue.employee]).Offer()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:62 +0x4a
  github.com/benz9527/xboot/lib/queue.TestArrayDelayQueue_PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_test.go:47 +0x70e
  testing.tRunner()
      C:/Program Files/Go/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      C:/Program Files/Go/src/testing/testing.go:1742 +0x44

Goroutine 9 (running) created at:
  github.com/benz9527/xboot/lib/queue.TestArrayDelayQueue_PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_test.go:38 +0x356
  testing.tRunner()
      C:/Program Files/Go/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      C:/Program Files/Go/src/testing/testing.go:1742 +0x44

Goroutine 8 (running) created at:
  testing.(*T).Run()
      C:/Program Files/Go/src/testing/testing.go:1742 +0x825
  testing.runTests.func1()
      C:/Program Files/Go/src/testing/testing.go:2161 +0x85
  testing.tRunner()
      C:/Program Files/Go/src/testing/testing.go:1689 +0x21e
  testing.runTests()
      C:/Program Files/Go/src/testing/testing.go:2159 +0x8be
  testing.(*M).Run()
      C:/Program Files/Go/src/testing/testing.go:2027 +0xf17
  main.main()
      _testmain.go:67 +0x2bd
==================
==================
WARNING: DATA RACE
Read at 0x00c00018c038 by goroutine 9:
  github.com/benz9527/xboot/lib/queue.(*ArrayPriorityQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).Peek()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:121 +0x54
  github.com/benz9527/xboot/lib/queue.(*ArrayPriorityQueue[*github.com/benz9527/xboot/lib/queue.employee]).Peek()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:120 +0x24
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).popIfExpired()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:49 +0x68
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).poll()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:92 +0x15a
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_windows.go:21 +0x144
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[*github.com/benz9527/xboot/lib/queue.employee]).PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_windows.go:11 +0x54
  github.com/benz9527/xboot/lib/queue.TestArrayDelayQueue_PollToChan.gowrap1()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_test.go:38 +0x6f

Previous write at 0x00c00018c038 by goroutine 8:
  github.com/benz9527/xboot/lib/queue.(*arrayPQ[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).Push()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:97 +0x1a7
  github.com/benz9527/xboot/lib/queue.(*arrayPQ[*github.com/benz9527/xboot/lib/queue.employee]).Push()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:88 +0x4a
  container/heap.Push()
      C:/Program Files/Go/src/container/heap/heap.go:52 +0x45
  github.com/benz9527/xboot/lib/queue.(*ArrayPriorityQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).Push()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:117 +0xa4
  github.com/benz9527/xboot/lib/queue.(*ArrayPriorityQueue[*github.com/benz9527/xboot/lib/queue.employee]).Push()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:116 +0x2e
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).Offer()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:65 +0x206
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[*github.com/benz9527/xboot/lib/queue.employee]).Offer()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:62 +0x4a
  github.com/benz9527/xboot/lib/queue.TestArrayDelayQueue_PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_test.go:50 +0x8cb
  testing.tRunner()
      C:/Program Files/Go/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      C:/Program Files/Go/src/testing/testing.go:1742 +0x44

Goroutine 9 (running) created at:
  github.com/benz9527/xboot/lib/queue.TestArrayDelayQueue_PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_test.go:38 +0x356
  testing.tRunner()
      C:/Program Files/Go/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      C:/Program Files/Go/src/testing/testing.go:1742 +0x44

Goroutine 8 (running) created at:
  testing.(*T).Run()
      C:/Program Files/Go/src/testing/testing.go:1742 +0x825
  testing.runTests.func1()
      C:/Program Files/Go/src/testing/testing.go:2161 +0x85
  testing.tRunner()
      C:/Program Files/Go/src/testing/testing.go:1689 +0x21e
  testing.runTests()
      C:/Program Files/Go/src/testing/testing.go:2159 +0x8be
  testing.(*M).Run()
      C:/Program Files/Go/src/testing/testing.go:2027 +0xf17
  main.main()
      _testmain.go:67 +0x2bd
==================
==================
WARNING: DATA RACE
Read at 0x00c000180070 by goroutine 9:
  github.com/benz9527/xboot/lib/queue.(*dqItem[*github.com/benz9527/xboot/lib/queue.employee]).Priority()
      <autogenerated>:1 +0x31
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).popIfExpired()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:51 +0x71
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).poll()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:92 +0x15a
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_windows.go:21 +0x144
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[*github.com/benz9527/xboot/lib/queue.employee]).PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_windows.go:11 +0x54
  github.com/benz9527/xboot/lib/queue.TestArrayDelayQueue_PollToChan.gowrap1()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_test.go:38 +0x6f

Previous write at 0x00c000180070 by goroutine 8:
  github.com/benz9527/xboot/lib/queue.NewDelayQueueItem[go.shape.*github.com/benz9527/xboot/lib/queue.employee]()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:23 +0x124
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).Offer()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:63 +0x36
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[*github.com/benz9527/xboot/lib/queue.employee]).Offer()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:62 +0x4a
  github.com/benz9527/xboot/lib/queue.TestArrayDelayQueue_PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_test.go:46 +0x675
  testing.tRunner()
      C:/Program Files/Go/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      C:/Program Files/Go/src/testing/testing.go:1742 +0x44

Goroutine 9 (running) created at:
  github.com/benz9527/xboot/lib/queue.TestArrayDelayQueue_PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_test.go:38 +0x356
  testing.tRunner()
      C:/Program Files/Go/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      C:/Program Files/Go/src/testing/testing.go:1742 +0x44

Goroutine 8 (running) created at:
  testing.(*T).Run()
      C:/Program Files/Go/src/testing/testing.go:1742 +0x825
  testing.runTests.func1()
      C:/Program Files/Go/src/testing/testing.go:2161 +0x85
  testing.tRunner()
      C:/Program Files/Go/src/testing/testing.go:1689 +0x21e
  testing.runTests()
      C:/Program Files/Go/src/testing/testing.go:2159 +0x8be
  testing.(*M).Run()
      C:/Program Files/Go/src/testing/testing.go:2027 +0xf17
  main.main()
      _testmain.go:67 +0x2bd
==================
==================
WARNING: DATA RACE
Read at 0x00c00018a048 by goroutine 9:
  sync/atomic.LoadInt64()
      C:/Program Files/Go/src/runtime/race_amd64.s:208 +0xb
  sync/atomic.LoadInt64()
      <autogenerated>:1 +0x10
  github.com/benz9527/xboot/lib/queue.(*dqItem[*github.com/benz9527/xboot/lib/queue.employee]).Priority()
      <autogenerated>:1 +0x43
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).popIfExpired()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:51 +0x71
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).poll()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:92 +0x15a
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_windows.go:21 +0x144
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[*github.com/benz9527/xboot/lib/queue.employee]).PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_windows.go:11 +0x54
  github.com/benz9527/xboot/lib/queue.TestArrayDelayQueue_PollToChan.gowrap1()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_test.go:38 +0x6f

Previous write at 0x00c00018a048 by goroutine 8:
  github.com/benz9527/xboot/lib/queue.NewPriorityQueueItem[go.shape.*github.com/benz9527/xboot/lib/queue.employee]()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:52 +0x7b
  github.com/benz9527/xboot/lib/queue.NewDelayQueueItem[go.shape.*github.com/benz9527/xboot/lib/queue.employee]()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:23 +0x52
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).Offer()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:63 +0x36
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[*github.com/benz9527/xboot/lib/queue.employee]).Offer()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:62 +0x4a
  github.com/benz9527/xboot/lib/queue.TestArrayDelayQueue_PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_test.go:46 +0x675
  testing.tRunner()
      C:/Program Files/Go/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      C:/Program Files/Go/src/testing/testing.go:1742 +0x44

Goroutine 9 (running) created at:
  github.com/benz9527/xboot/lib/queue.TestArrayDelayQueue_PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_test.go:38 +0x356
  testing.tRunner()
      C:/Program Files/Go/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      C:/Program Files/Go/src/testing/testing.go:1742 +0x44

Goroutine 8 (running) created at:
  testing.(*T).Run()
      C:/Program Files/Go/src/testing/testing.go:1742 +0x825
  testing.runTests.func1()
      C:/Program Files/Go/src/testing/testing.go:2161 +0x85
  testing.tRunner()
      C:/Program Files/Go/src/testing/testing.go:1689 +0x21e
  testing.runTests()
      C:/Program Files/Go/src/testing/testing.go:2159 +0x8be
  testing.(*M).Run()
      C:/Program Files/Go/src/testing/testing.go:2027 +0xf17
  main.main()
      _testmain.go:67 +0x2bd
==================
==================
WARNING: DATA RACE
Write at 0x00c00019e080 by goroutine 8:
  github.com/benz9527/xboot/lib/queue.(*arrayPQ[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).Swap()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:70 +0x164
  github.com/benz9527/xboot/lib/queue.(*arrayPQ[*github.com/benz9527/xboot/lib/queue.employee]).Swap()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:69 +0x4a
  container/heap.up()
      C:/Program Files/Go/src/container/heap/heap.go:95 +0x4b
  container/heap.Push()
      C:/Program Files/Go/src/container/heap/heap.go:53 +0x68
  github.com/benz9527/xboot/lib/queue.(*ArrayPriorityQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).Push()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:117 +0xa4
  github.com/benz9527/xboot/lib/queue.(*ArrayPriorityQueue[*github.com/benz9527/xboot/lib/queue.employee]).Push()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:116 +0x2e
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).Offer()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:65 +0x206
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[*github.com/benz9527/xboot/lib/queue.employee]).Offer()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:62 +0x4a
  github.com/benz9527/xboot/lib/queue.TestArrayDelayQueue_PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_test.go:50 +0x8cb
  testing.tRunner()
      C:/Program Files/Go/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      C:/Program Files/Go/src/testing/testing.go:1742 +0x44

Previous read at 0x00c00019e080 by goroutine 9:
  github.com/benz9527/xboot/lib/queue.(*ArrayPriorityQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).Peek()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:124 +0xa9
  github.com/benz9527/xboot/lib/queue.(*ArrayPriorityQueue[*github.com/benz9527/xboot/lib/queue.employee]).Peek()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:120 +0x24
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).popIfExpired()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:49 +0x68
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).poll()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:92 +0x15a
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_windows.go:21 +0x144
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[*github.com/benz9527/xboot/lib/queue.employee]).PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_windows.go:11 +0x54
  github.com/benz9527/xboot/lib/queue.TestArrayDelayQueue_PollToChan.gowrap1()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_test.go:38 +0x6f

Goroutine 8 (running) created at:
  testing.(*T).Run()
      C:/Program Files/Go/src/testing/testing.go:1742 +0x825
  testing.runTests.func1()
      C:/Program Files/Go/src/testing/testing.go:2161 +0x85
  testing.tRunner()
      C:/Program Files/Go/src/testing/testing.go:1689 +0x21e
  testing.runTests()
      C:/Program Files/Go/src/testing/testing.go:2159 +0x8be
  testing.(*M).Run()
      C:/Program Files/Go/src/testing/testing.go:2027 +0xf17
  main.main()
      _testmain.go:67 +0x2bd

Goroutine 9 (running) created at:
  github.com/benz9527/xboot/lib/queue.TestArrayDelayQueue_PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_test.go:38 +0x356
  testing.tRunner()
      C:/Program Files/Go/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      C:/Program Files/Go/src/testing/testing.go:1742 +0x44
==================
==================
WARNING: DATA RACE
Write at 0x00c00018c038 by goroutine 8:
  github.com/benz9527/xboot/lib/queue.(*arrayPQ[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).Push()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:97 +0x1a7
  github.com/benz9527/xboot/lib/queue.(*arrayPQ[*github.com/benz9527/xboot/lib/queue.employee]).Push()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:88 +0x4a
  container/heap.Push()
      C:/Program Files/Go/src/container/heap/heap.go:52 +0x45
  github.com/benz9527/xboot/lib/queue.(*ArrayPriorityQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).Push()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:117 +0xa4
  github.com/benz9527/xboot/lib/queue.(*ArrayPriorityQueue[*github.com/benz9527/xboot/lib/queue.employee]).Push()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:116 +0x2e
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).Offer()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:65 +0x206
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[*github.com/benz9527/xboot/lib/queue.employee]).Offer()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:62 +0x4a
  github.com/benz9527/xboot/lib/queue.TestArrayDelayQueue_PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_test.go:51 +0x962
  testing.tRunner()
      C:/Program Files/Go/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      C:/Program Files/Go/src/testing/testing.go:1742 +0x44

Previous read at 0x00c00018c038 by goroutine 9:
  github.com/benz9527/xboot/lib/queue.(*ArrayPriorityQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).Len()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:105 +0x4c
  github.com/benz9527/xboot/lib/queue.(*ArrayPriorityQueue[*github.com/benz9527/xboot/lib/queue.employee]).Len()
      D:/workspace/github.com/benz9527/xboot/lib/queue/priority_queue.go:104 +0x17
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).popIfExpired()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:45 +0x43
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).poll()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue.go:92 +0x15a
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[go.shape.*github.com/benz9527/xboot/lib/queue.employee]).PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_windows.go:21 +0x144
  github.com/benz9527/xboot/lib/queue.(*ArrayDelayQueue[*github.com/benz9527/xboot/lib/queue.employee]).PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_windows.go:11 +0x54
  github.com/benz9527/xboot/lib/queue.TestArrayDelayQueue_PollToChan.gowrap1()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_test.go:38 +0x6f

Goroutine 8 (running) created at:
  testing.(*T).Run()
      C:/Program Files/Go/src/testing/testing.go:1742 +0x825
  testing.runTests.func1()
      C:/Program Files/Go/src/testing/testing.go:2161 +0x85
  testing.tRunner()
      C:/Program Files/Go/src/testing/testing.go:1689 +0x21e
  testing.runTests()
      C:/Program Files/Go/src/testing/testing.go:2159 +0x8be
  testing.(*M).Run()
      C:/Program Files/Go/src/testing/testing.go:2027 +0xf17
  main.main()
      _testmain.go:67 +0x2bd

Goroutine 9 (running) created at:
  github.com/benz9527/xboot/lib/queue.TestArrayDelayQueue_PollToChan()
      D:/workspace/github.com/benz9527/xboot/lib/queue/delay_queue_test.go:38 +0x356
  testing.tRunner()
      C:/Program Files/Go/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      C:/Program Files/Go/src/testing/testing.go:1742 +0x44
==================
--- FAIL: TestArrayDelayQueue_PollToChan (0.30s)
    delay_queue_test.go:78: current time ms: 1708676038321, item: &{p4 3 1708676038320}, diff: 1
    delay_queue_test.go:78: current time ms: 1708676038370, item: &{p0 10 1708676038370}, diff: 0
    delay_queue_test.go:78: current time ms: 1708676038370, item: &{p5 1 1708676038370}, diff: 0
    delay_queue_test.go:78: current time ms: 1708676038415, item: &{p2 10 1708676038415}, diff: 0
    delay_queue_test.go:78: current time ms: 1708676038470, item: &{p3 200 1708676038470}, diff: 0
    delay_queue_test.go:78: current time ms: 1708676038511, item: &{p6 5 1708676038510}, diff: 1
    delay_queue_test.go:78: current time ms: 1708676038562, item: &{p7 200 1708676038561}, diff: 1
    testing.go:1398: race detected during execution of test
FAIL
FAIL	github.com/benz9527/xboot/lib/queue	0.629s
FAIL
```

--- a/lib/list/linkedlist.go
+++ b/lib/list/linkedlist.go
@@ -269,6 +269,9 @@ func (l *doublyLinkedList[T]) Remove(targetE NodeElement[T]) NodeElement[T] {
 }
 
 func (l *doublyLinkedList[T]) ForEach(fn func(idx int64, e NodeElement[T])) {
+	if l == nil {
+		return
+	}
 	if fn == nil || l.len.Load() == 0 ||
 		l.getRoot() == l.getRootHead() && l.getRoot() == l.getRootTail() {
 		return
@@ -287,6 +290,9 @@ func (l *doublyLinkedList[T]) ForEach(fn func(idx int64, e NodeElement[T])) {
 }
 
 func (l *doublyLinkedList[T]) ReverseForEach(fn func(idx int64, e NodeElement[T])) {
+	if l == nil {
+		return
+	}
 	if fn == nil || l.len.Load() == 0 ||
 		l.getRoot() == l.getRootHead() && l.getRoot() == l.getRootTail() {
 		return
@@ -305,7 +311,7 @@ func (l *doublyLinkedList[T]) ReverseForEach(fn func(idx int64, e NodeElement[T]
 }
 
 func (l *doublyLinkedList[T]) FindFirst(targetV T, compareFn ...func(e NodeElement[T]) bool) (NodeElement[T], bool) {
-	if l.len.Load() == 0 {
+	if l == nil || l.len.Load() == 0 {
 		return nil, false
 	}
 
@@ -328,18 +334,30 @@ func (l *doublyLinkedList[T]) FindFirst(targetV T, compareFn ...func(e NodeEleme
 }
 
 func (l *doublyLinkedList[T]) Front() NodeElement[T] {
+	if l == nil {
+		return nil
+	}
 	return l.root.GetNext()
 }
 
 func (l *doublyLinkedList[T]) Back() NodeElement[T] {
+	if l == nil {
+		return nil
+	}
 	return l.root.GetPrev()
 }
 
 func (l *doublyLinkedList[T]) PushFront(v T) NodeElement[T] {
+	if l == nil {
+		return nil
+	}
 	return l.InsertBefore(v, l.getRootHead())
 }
 
 func (l *doublyLinkedList[T]) PushBack(v T) NodeElement[T] {
+	if l == nil {
+		return nil
+	}
 	return l.InsertAfter(v, l.getRootTail())
 }
 
@@ -370,6 +388,9 @@ func (l *doublyLinkedList[T]) move(src, dst *nodeElement[T]) bool {
 }
 
 func (l *doublyLinkedList[T]) MoveToFront(targetE NodeElement[T]) {
+	if l == nil {
+		return
+	}
 	moved := false
 	src, status := l.checkElement(targetE)
 	switch status {
@@ -389,6 +410,9 @@ func (l *doublyLinkedList[T]) MoveToFront(targetE NodeElement[T]) {
 }
 
 func (l *doublyLinkedList[T]) MoveToBack(targetE NodeElement[T]) {
+	if l == nil {
+		return
+	}
 	moved := false
 	src, status := l.checkElement(targetE)
 	switch status {
@@ -408,6 +432,9 @@ func (l *doublyLinkedList[T]) MoveToBack(targetE NodeElement[T]) {
 }
 
 func (l *doublyLinkedList[T]) MoveBefore(srcE, dstE NodeElement[T]) {
+	if l == nil {
+		return
+	}
 	var (
 		moved                = false
 		dst, src             *nodeElement[T]
@@ -458,6 +485,9 @@ func (l *doublyLinkedList[T]) MoveBefore(srcE, dstE NodeElement[T]) {
 }
 
 func (l *doublyLinkedList[T]) MoveAfter(srcE, dstE NodeElement[T]) {
+	if l == nil {
+		return
+	}
 	var (
 		moved                = false
 		dst, src             *nodeElement[T]
@@ -505,6 +535,9 @@ func (l *doublyLinkedList[T]) MoveAfter(srcE, dstE NodeElement[T]) {
 }
 
 func (l *doublyLinkedList[T]) PushFrontList(src LinkedList[T]) {
+	if l == nil {
+		return
+	}
 	if dl, ok := src.(*doublyLinkedList[T]); !ok || dl != nil && dl.getRoot() == l.getRoot() {
 		// avoid type mismatch and self copy
 		return
@@ -522,6 +555,9 @@ func (l *doublyLinkedList[T]) PushFrontList(src LinkedList[T]) {
 }
 
 func (l *doublyLinkedList[T]) PushBackList(src LinkedList[T]) {
+	if l == nil {
+		return
+	}
 	if dl, ok := src.(*doublyLinkedList[T]); !ok || dl != nil && dl.getRoot() == l.getRoot() {
 		// avoid type mismatch and self copy
 		return

--- a/lib/queue/delay_queue.go
+++ b/lib/queue/delay_queue.go
@@ -33,6 +33,7 @@ const (
 
 type ArrayDelayQueue[E comparable] struct {
 	pq                  PriorityQueue[E]
+	itemCounter         atomic.Int64
 	workCtx             context.Context
 	lock                *sync.Mutex
 	exclusion           *sync.Mutex // Avoid multiple poll
@@ -70,6 +71,7 @@ func (dq *ArrayDelayQueue[E]) Offer(item E, expiration int64) {
 			dq.wakeUpC <- struct{}{}
 		}
 	}
+	dq.itemCounter.Add(1)
 }
 
 func (dq *ArrayDelayQueue[E]) poll(nowFn func() int64, sender ipc.SendOnlyChannel[E]) {
@@ -151,11 +153,16 @@ func (dq *ArrayDelayQueue[E]) poll(nowFn func() int64, sender ipc.SendOnlyChanne
 				if err := sender.Send(item.Value()); err != nil {
 					return
 				}
+				dq.itemCounter.Add(-1)
 			} else {
 				return
 			}
 		}
 	}
+}
+
+func (dq *ArrayDelayQueue[E]) Len() int64 {
+	return dq.itemCounter.Load()
 }
 
 func NewArrayDelayQueue[E comparable](ctx context.Context, capacity int) DelayQueue[E] {

--- a/lib/queue/delay_queue.go
+++ b/lib/queue/delay_queue.go
@@ -170,7 +170,10 @@ func (dq *ArrayDelayQueue[E]) Len() int64 {
 
 func NewArrayDelayQueue[E comparable](ctx context.Context, capacity int) DelayQueue[E] {
 	dq := &ArrayDelayQueue[E]{
-		pq:                  NewArrayPriorityQueue[E](WithArrayPriorityQueueCapacity[E](capacity)),
+		pq: NewArrayPriorityQueue[E](
+			WithArrayPriorityQueueEnableThreadSafe[E](),
+			WithArrayPriorityQueueCapacity[E](capacity),
+		),
 		workCtx:             ctx,
 		lock:                &sync.Mutex{},
 		exclusion:           &sync.Mutex{},

--- a/lib/queue/delay_queue.go
+++ b/lib/queue/delay_queue.go
@@ -162,6 +162,9 @@ func (dq *ArrayDelayQueue[E]) poll(nowFn func() int64, sender ipc.SendOnlyChanne
 }
 
 func (dq *ArrayDelayQueue[E]) Len() int64 {
+	if dq == nil {
+		return 0
+	}
 	return dq.itemCounter.Load()
 }
 

--- a/lib/queue/priority_queue.go
+++ b/lib/queue/priority_queue.go
@@ -2,6 +2,7 @@ package queue
 
 import (
 	"container/heap"
+	"sync"
 	"sync/atomic"
 )
 
@@ -99,13 +100,22 @@ func (pq *arrayPQ[E]) Push(i interface{}) {
 
 type ArrayPriorityQueue[E comparable] struct {
 	queue *arrayPQ[E]
+	lock  *sync.Mutex
 }
 
 func (pq *ArrayPriorityQueue[E]) Len() int64 {
+	if pq.lock != nil {
+		pq.lock.Lock()
+		defer pq.lock.Unlock()
+	}
 	return int64(len(pq.queue.arr))
 }
 
 func (pq *ArrayPriorityQueue[E]) Pop() (nilItem ReadOnlyPQItem[E]) {
+	if pq.lock != nil {
+		pq.lock.Lock()
+		defer pq.lock.Unlock()
+	}
 	if len(pq.queue.arr) == 0 {
 		return nil
 	}
@@ -114,10 +124,18 @@ func (pq *ArrayPriorityQueue[E]) Pop() (nilItem ReadOnlyPQItem[E]) {
 }
 
 func (pq *ArrayPriorityQueue[E]) Push(item PQItem[E]) {
+	if pq.lock != nil {
+		pq.lock.Lock()
+		defer pq.lock.Unlock()
+	}
 	heap.Push(pq.queue, item)
 }
 
 func (pq *ArrayPriorityQueue[E]) Peek() ReadOnlyPQItem[E] {
+	if pq.lock != nil {
+		pq.lock.Lock()
+		defer pq.lock.Unlock()
+	}
 	if len(pq.queue.arr) == 0 {
 		return nil
 	}
@@ -175,5 +193,11 @@ func WithArrayPriorityQueueComparator[E comparable](fn PQItemLessThenComparator[
 			}
 		}
 		pq.queue.comparator = fn
+	}
+}
+
+func WithArrayPriorityQueueEnableThreadSafe[E comparable]() ArrayPriorityQueueOption[E] {
+	return func(pq *ArrayPriorityQueue[E]) {
+		pq.lock = &sync.Mutex{}
 	}
 }

--- a/lib/queue/priority_queue_test.go
+++ b/lib/queue/priority_queue_test.go
@@ -30,6 +30,7 @@ func TestPriorityQueueItemAlignmentAndSize(t *testing.T) {
 
 func TestPriorityQueue_MinValueAsHighPriority(t *testing.T) {
 	pq := NewArrayPriorityQueue[*employee](
+		WithArrayPriorityQueueEnableThreadSafe[*employee](),
 		WithArrayPriorityQueueCapacity[*employee](32),
 		WithArrayPriorityQueueComparator[*employee](func(i, j ReadOnlyPQItem[*employee]) CmpEnum {
 			res := i.Priority() - j.Priority()
@@ -59,6 +60,7 @@ func TestPriorityQueue_MinValueAsHighPriority(t *testing.T) {
 
 func TestPriorityQueue_MaxValueAsHighPriority(t *testing.T) {
 	pq := NewArrayPriorityQueue[*employee](
+		WithArrayPriorityQueueEnableThreadSafe[*employee](),
 		WithArrayPriorityQueueCapacity[*employee](32),
 		WithArrayPriorityQueueComparator[*employee](func(i, j ReadOnlyPQItem[*employee]) CmpEnum {
 			res := j.Priority() - i.Priority()
@@ -89,6 +91,7 @@ func TestPriorityQueue_MaxValueAsHighPriority(t *testing.T) {
 
 func TestPriorityQueue_MinValueAsHighPriority_Peek(t *testing.T) {
 	pq := NewArrayPriorityQueue[*employee](
+		WithArrayPriorityQueueEnableThreadSafe[*employee](),
 		WithArrayPriorityQueueCapacity[*employee](32),
 		WithArrayPriorityQueueComparator[*employee](func(i, j ReadOnlyPQItem[*employee]) CmpEnum {
 			res := i.Priority() - j.Priority()
@@ -121,6 +124,7 @@ func TestPriorityQueue_MinValueAsHighPriority_Peek(t *testing.T) {
 
 func TestPriorityQueue_MaxValueAsHighPriority_Peek(t *testing.T) {
 	pq := NewArrayPriorityQueue[*employee](
+		WithArrayPriorityQueueEnableThreadSafe[*employee](),
 		WithArrayPriorityQueueCapacity[*employee](32),
 		WithArrayPriorityQueueComparator[*employee](func(i, j ReadOnlyPQItem[*employee]) CmpEnum {
 			res := j.Priority() - i.Priority()
@@ -159,6 +163,7 @@ func BenchmarkArrayPriorityQueue_Push(b *testing.B) {
 	}
 	b.ResetTimer()
 	pq := NewArrayPriorityQueue[*employee](
+		WithArrayPriorityQueueEnableThreadSafe[*employee](),
 		WithArrayPriorityQueueCapacity[*employee](32),
 		WithArrayPriorityQueueComparator[*employee](func(i, j ReadOnlyPQItem[*employee]) CmpEnum {
 			res := j.Priority() - i.Priority()
@@ -182,7 +187,10 @@ func BenchmarkArrayPriorityQueue_Pop(b *testing.B) {
 		e := NewPriorityQueueItem[*employee](&employee{age: i, name: fmt.Sprintf("p%d", i)}, int64(i))
 		list = append(list, e)
 	}
-	pq := NewArrayPriorityQueue[*employee](WithArrayPriorityQueueCapacity[*employee](32))
+	pq := NewArrayPriorityQueue[*employee](
+		WithArrayPriorityQueueEnableThreadSafe[*employee](),
+		WithArrayPriorityQueueCapacity[*employee](32),
+	)
 	for i := 0; i < b.N; i++ {
 		pq.Push(list[i])
 	}

--- a/lib/queue/queue_intf.go
+++ b/lib/queue/queue_intf.go
@@ -43,6 +43,7 @@ type DelayQueue[E comparable] interface {
 	Offer(item E, expiration int64)
 	// PollToChan Asynchronous function
 	PollToChan(nowFn func() int64, C ipc.SendOnlyChannel[E])
+	Len() int64
 }
 
 type DQItem[E comparable] interface {

--- a/timer/stats.go
+++ b/timer/stats.go
@@ -17,38 +17,56 @@ const (
 )
 
 type timingWheelStats struct {
-	minTickMs             int64
-	clock                 hrtime.Clock
-	jobExecutedCounter    atomic.Int64
-	jobHighLatencyCounter atomic.Int64
-	jobCount              metric.Int64UpDownCounter
-	jobTickAccuracy       metric.Float64ObservableGauge
-	jobLatencies          metric.Int64Histogram
-	jobExecuteDurations   metric.Int64Histogram
-	jobExecutedCount      metric.Int64Counter
-	jobCancelledCount     metric.Int64Counter
+	ctx                 context.Context
+	minTickMs           int64
+	clock               hrtime.Clock
+	jobExecutedCount    atomic.Int64
+	jobHighLatencyCount atomic.Int64
+	slotActiveCount     atomic.Int64
+	jobAliveCounter     metric.Int64UpDownCounter
+	jobTickAccuracy     metric.Float64ObservableGauge
+	jobLatencies        metric.Int64Histogram
+	jobExecuteDurations metric.Int64Histogram
+	jobExecutedCounter  metric.Int64Counter
+	jobCancelledCounter metric.Int64Counter
+	slotCounter         metric.Int64Counter
+	slotActiveCounter   metric.Int64ObservableUpDownCounter
 }
 
-func (stats *timingWheelStats) RecordJobCount(count int64) {
+func (stats *timingWheelStats) RecordJobAliveCount(count int64) {
 	if stats == nil {
 		return
 	}
-	stats.jobCount.Add(context.Background(), count)
+	stats.jobAliveCounter.Add(stats.ctx, count)
+}
+
+func (stats *timingWheelStats) UpdateSlotActiveCount(count int64) {
+	if stats == nil {
+		return
+	}
+	stats.slotActiveCount.Swap(count)
+}
+
+func (stats *timingWheelStats) RecordSlotCount(count int64) {
+	if stats == nil {
+		return
+	}
+	stats.slotCounter.Add(stats.ctx, count)
 }
 
 func (stats *timingWheelStats) IncreaseJobExecutedCount() {
 	if stats == nil {
 		return
 	}
-	stats.jobExecutedCount.Add(context.Background(), 1)
-	stats.jobExecutedCounter.Add(1)
+	stats.jobExecutedCounter.Add(stats.ctx, 1)
+	stats.jobExecutedCount.Add(1)
 }
 
 func (stats *timingWheelStats) IncreaseJobCancelledCount() {
 	if stats == nil {
 		return
 	}
-	stats.jobCancelledCount.Add(context.Background(), 1)
+	stats.jobCancelledCounter.Add(stats.ctx, 1)
 }
 
 func (stats *timingWheelStats) RecordJobLatency(latencyMs int64) {
@@ -58,9 +76,9 @@ func (stats *timingWheelStats) RecordJobLatency(latencyMs int64) {
 	as := attribute.NewSet(
 		attribute.String("xtw.job.latency.ms", strconv.FormatInt(latencyMs, 10)),
 	)
-	stats.jobLatencies.Record(context.Background(), 1, metric.WithAttributeSet(as))
+	stats.jobLatencies.Record(stats.ctx, 1, metric.WithAttributeSet(as))
 	if latencyMs > stats.minTickMs || latencyMs < -stats.minTickMs {
-		stats.jobHighLatencyCounter.Add(1)
+		stats.jobHighLatencyCount.Add(1)
 	}
 }
 
@@ -71,7 +89,7 @@ func (stats *timingWheelStats) RecordJobExecuteDuration(durationMs int64) {
 	as := attribute.NewSet(
 		attribute.String("xtw.job.execute.duration.ms", strconv.FormatInt(durationMs, 10)),
 	)
-	stats.jobExecuteDurations.Record(context.Background(), durationMs, metric.WithAttributeSet(as))
+	stats.jobExecuteDurations.Record(stats.ctx, durationMs, metric.WithAttributeSet(as))
 }
 
 func WithTimingWheelStats() TimingWheelsOption {
@@ -84,9 +102,10 @@ func newTimingWheelStats(ref *xTimingWheels) *timingWheelStats {
 	meterName := fmt.Sprintf("%s/%s", TimingWheelStatsName, ref.name)
 	tickMs := ref.GetTickMs()
 	stats := &timingWheelStats{
+		ctx:       context.Background(),
 		minTickMs: tickMs,
 		clock:     ref.clock,
-		jobCount: lo.Must[metric.Int64UpDownCounter](otel.Meter(meterName).
+		jobAliveCounter: lo.Must[metric.Int64UpDownCounter](otel.Meter(meterName).
 			Int64UpDownCounter(
 				"xtw.job.count",
 				metric.WithDescription("The number of jobs in the timing wheel."),
@@ -106,16 +125,22 @@ func newTimingWheelStats(ref *xTimingWheels) *timingWheelStats {
 				metric.WithUnit("ms"),
 			),
 		),
-		jobExecutedCount: lo.Must[metric.Int64Counter](otel.Meter(meterName).
+		jobExecutedCounter: lo.Must[metric.Int64Counter](otel.Meter(meterName).
 			Int64Counter(
 				"xtw.job.executed.count",
 				metric.WithDescription("The number of jobs executed by the timing wheel."),
 			),
 		),
-		jobCancelledCount: lo.Must[metric.Int64Counter](otel.Meter(meterName).
+		jobCancelledCounter: lo.Must[metric.Int64Counter](otel.Meter(meterName).
 			Int64Counter(
 				"xtw.job.cancelled.count",
 				metric.WithDescription("The number of jobs cancelled by the timing wheel."),
+			),
+		),
+		slotCounter: lo.Must[metric.Int64Counter](otel.Meter(meterName).
+			Int64Counter(
+				"xtw.slot.count",
+				metric.WithDescription("The number of slots belongs to the timing wheel."),
 			),
 		),
 	}
@@ -125,14 +150,25 @@ func newTimingWheelStats(ref *xTimingWheels) *timingWheelStats {
 			metric.WithDescription(fmt.Sprintf("The accuracy of the timing wheel tick [-%d,%d] ms.", tickMs, tickMs)),
 			metric.WithFloat64Callback(func(ctx context.Context, ob metric.Float64Observer) error {
 				accuracy := 1.00
-				if stats.jobExecutedCounter.Load() > 0 {
-					accuracy = float64(stats.jobExecutedCounter.Load()-stats.jobHighLatencyCounter.Load()) /
-						float64(stats.jobExecutedCounter.Load())
+				if stats.jobExecutedCount.Load() > 0 {
+					accuracy = float64(stats.jobExecutedCount.Load()-stats.jobHighLatencyCount.Load()) /
+						float64(stats.jobExecutedCount.Load())
 				}
 				ob.Observe(accuracy)
 				return nil
 			}),
 			metric.WithUnit("%"),
+		),
+	)
+	stats.slotActiveCounter = lo.Must[metric.Int64ObservableUpDownCounter](otel.Meter(meterName).
+		Int64ObservableUpDownCounter(
+			"xtw.slot.active.count",
+			metric.WithDescription("The number of slots in active (expired) belongs to the timing wheel."),
+			metric.WithInt64Callback(func(ctx context.Context, ob metric.Int64Observer) error {
+				slots := stats.slotActiveCount.Load()
+				ob.Observe(slots)
+				return nil
+			}),
 		),
 	)
 	return stats

--- a/timer/stats_windows.go
+++ b/timer/stats_windows.go
@@ -5,7 +5,6 @@ package timer
 
 import (
 	"context"
-	"time"
 )
 
 func jobStatsWrapper(stats *timingWheelStats, invoke Job) Job {
@@ -13,9 +12,7 @@ func jobStatsWrapper(stats *timingWheelStats, invoke Job) Job {
 		return invoke
 	}
 	return func(ctx context.Context, metadata JobMetadata) {
-		var beginTime time.Time
-		beginTime = stats.clock.NowInDefaultTZ()
-
+		var beginTime = stats.clock.NowInDefaultTZ()
 		defer func() {
 			stats.IncreaseJobExecutedCount()
 			stats.RecordJobExecuteDuration(stats.clock.Since(beginTime).Milliseconds())

--- a/timer/timing_wheel_intf.go
+++ b/timer/timing_wheel_intf.go
@@ -29,14 +29,10 @@ var (
 type TimingWheelCommonMetadata interface {
 	// GetTickMs returns the baseline tick ms (interval) of the timing wheel.
 	GetTickMs() int64
-	// GetSlotSize returns the slot size of the timing wheel.
-	GetSlotSize() int64
 	// GetStartMs returns the start ms of the timing wheel.
 	GetStartMs() int64
 }
 
-// TimingWheel slots is private,
-// they should be provided by the implementation
 type TimingWheel interface {
 	TimingWheelCommonMetadata
 	GetInterval() int64
@@ -53,8 +49,6 @@ type Scheduler interface {
 
 type TimingWheels interface {
 	TimingWheelCommonMetadata
-	// GetTaskCounter returns the current task count of the timing wheel.
-	GetTaskCounter() int64
 	// AddTask adds a task to the timing wheels.
 	AddTask(task Task) error
 	// CancelTask cancels a task by jobID.

--- a/timer/x_slot.go
+++ b/timer/x_slot.go
@@ -1,6 +1,7 @@
 package timer
 
 import (
+	"sync"
 	"sync/atomic"
 
 	"github.com/benz9527/xboot/lib/list"
@@ -41,6 +42,7 @@ func (slot *slotMetadata) setLevel(level int64) {
 // xSlot the segment rwlock free
 type xSlot struct {
 	*slotMetadata
+	lock  *sync.Mutex
 	tasks list.LinkedList[Task]
 }
 
@@ -60,6 +62,7 @@ func NewXSlot() TimingWheelSlot {
 		slotMetadata: &slotMetadata{
 			expirationMs: -1,
 		},
+		lock:  &sync.Mutex{},
 		tasks: list.NewLinkedList[Task](),
 	}
 }
@@ -73,11 +76,15 @@ func newSentinelSlot() TimingWheelSlot {
 }
 
 func (slot *xSlot) GetMetadata() TimingWheelSlotMetadata {
+	slot.lock.Lock()
+	defer slot.lock.Unlock()
 	metadata := *slot.slotMetadata // Copy instead of reference
 	return &metadata
 }
 
 func (slot *xSlot) AddTask(task Task) {
+	slot.lock.Lock()
+	defer slot.lock.Unlock()
 	if task == nil && slot.GetExpirationMs() == slotHasBeenFlushedMs {
 		return
 	}
@@ -111,6 +118,8 @@ func (slot *xSlot) removeTask(task Task) bool {
 
 // RemoveTask the slot must be not expired.
 func (slot *xSlot) RemoveTask(task Task) bool {
+	slot.lock.Lock()
+	defer slot.lock.Unlock()
 	if slot.GetExpirationMs() == slotHasBeenFlushedMs {
 		return false
 	}
@@ -120,6 +129,8 @@ func (slot *xSlot) RemoveTask(task Task) bool {
 
 // Flush Timing-wheel scheduling algorithm core function
 func (slot *xSlot) Flush(reinsert TaskHandler) {
+	slot.lock.Lock()
+	defer slot.lock.Unlock()
 	if slot == nil || slot.tasks == nil {
 		return
 	}

--- a/timer/x_slot.go
+++ b/timer/x_slot.go
@@ -120,6 +120,9 @@ func (slot *xSlot) RemoveTask(task Task) bool {
 
 // Flush Timing-wheel scheduling algorithm core function
 func (slot *xSlot) Flush(reinsert TaskHandler) {
+	if slot == nil || slot.tasks == nil {
+		return
+	}
 	// Reset the slot, ready for next round.
 	slot.setExpirationMs(slotHasBeenFlushedMs)
 	// Due to the slot has been expired, we have to handle all tasks from the slot.
@@ -133,6 +136,7 @@ func (slot *xSlot) Flush(reinsert TaskHandler) {
 	//      Otherwise, cancel it.
 	// 3. Remove the tasks from the slot.
 	// 4. Reset the slot, ready for next round.
+	// Not the atomic operation, it is possible to be nil still
 	slot.tasks.ForEach(func(idx int64, iterator list.NodeElement[Task]) {
 		task := iterator.GetValue()
 		slot.removeTask(task) // clean task reference at first

--- a/timer/x_timing_wheels.go
+++ b/timer/x_timing_wheels.go
@@ -109,18 +109,14 @@ func (tw *timingWheel) addTask(task Task, level int64) error {
 		virtualID := taskExpiredMs / tickMs
 		slotID := virtualID % slotSize
 		slot := tw.slots[slotID]
-		slotMs := slot.GetExpirationMs()
-		if slot.GetExpirationMs() != (virtualID*tickMs) && !slot.setExpirationMs(virtualID*tickMs) { // FIXME data race
-			err := fmt.Errorf("[timing wheel] slot (level:%d) (old:%d<->new:%d) unable update the expiration, %w",
-				level, slotMs, virtualID*tickMs, ErrTimingWheelTaskUnableToBeAddedToSlot)
-			slog.Error("[timing wheel] add task error", "error", err)
-			return err
+		if slot.GetExpirationMs() == virtualID*tickMs {
+			slot.AddTask(task)
+		} else if slot.setExpirationMs(virtualID * tickMs) {
+			slot.setSlotID(slotID)
+			slot.setLevel(level)
+			slot.AddTask(task)
+			tw.globalDqRef.Offer(slot, slot.GetExpirationMs())
 		}
-
-		slot.setSlotID(slotID)
-		slot.setLevel(level)
-		slot.AddTask(task)
-		tw.globalDqRef.Offer(slot, slot.GetExpirationMs())
 		return nil
 	}
 	// Out of the interval. Put it into the higher interval timing wheel
@@ -204,20 +200,24 @@ func (xtw *xTimingWheels) GetStartMs() int64 {
 }
 
 func (xtw *xTimingWheels) Shutdown() {
+	if xtw == nil {
+		return
+	}
 	if old := xtw.isRunning.Swap(false); !old {
 		slog.Warn("[timing wheel] timing wheel is already shutdown")
 		return
 	}
-	xtw.dq = nil
 	xtw.isRunning.Store(false)
 
-	// FIXME close on channel is no empty and will cause panic.
 	close(xtw.stopC)
 	_ = xtw.expiredSlotC.Close()
 	_ = xtw.twEventC.Close()
 	xtw.gPool.Release()
 
-	// FIXME map clear data race
+	runtime.SetFinalizer(xtw, func(xtw *xTimingWheels) {
+		xtw.dq = nil
+		clear(xtw.tasksMap)
+	})
 }
 
 func (xtw *xTimingWheels) AddTask(task Task) error {

--- a/timer/x_timing_wheels.go
+++ b/timer/x_timing_wheels.go
@@ -27,16 +27,16 @@ type timingWheel struct {
 	slots []TimingWheelSlot // alignment 8, size 24; in kafka it is buckets
 	// ctx is used to shut down the timing wheel and pass
 	// value to control debug info.
-	ctx                  context.Context
-	globalDqRef          queue.DelayQueue[TimingWheelSlot]
-	overflowWheelRef     unsafe.Pointer // same as kafka TimingWheel(*timingWheel)
-	tickMs               int64
-	startMs              int64 // baseline startup timestamp
-	interval             int64
-	currentTimeMs        int64
-	slotSize             int64 // in kafka it is wheelSize
-	globalSlotCounterRef *atomic.Int64
-	clock                hrtime.Clock
+	ctx              context.Context
+	globalDqRef      queue.DelayQueue[TimingWheelSlot]
+	overflowWheelRef unsafe.Pointer // same as kafka TimingWheel(*timingWheel)
+	tickMs           int64
+	startMs          int64 // baseline startup timestamp
+	interval         int64
+	currentTimeMs    int64
+	slotSize         int64 // in kafka it is wheelSize
+	globalStats      *timingWheelStats
+	clock            hrtime.Clock
 }
 
 func (tw *timingWheel) GetTickMs() int64 {
@@ -103,7 +103,6 @@ func (tw *timingWheel) addTask(task Task, level int64) error {
 	if diff < tickMs {
 		task.setSlot(immediateExpiredSlot)
 		tw.advanceClock(currentTimeMs)
-		tw.globalSlotCounterRef.Add(1)
 		return fmt.Errorf("[timing wheel] task task expired ms  %d is before %d, %w",
 			taskExpiredMs, currentTimeMs+tickMs, ErrTimingWheelTaskIsExpired)
 	} else if diff >= tickMs && diff < interval {
@@ -132,7 +131,7 @@ func (tw *timingWheel) addTask(task Task, level int64) error {
 			interval,
 			slotSize,
 			currentTimeMs,
-			tw.globalSlotCounterRef,
+			tw.globalStats,
 			tw.globalDqRef,
 			tw.clock,
 		))
@@ -146,27 +145,27 @@ func newTimingWheel(
 	tickMs int64,
 	slotSize int64,
 	startMs int64,
-	slotCounter *atomic.Int64,
+	stats *timingWheelStats,
 	dq queue.DelayQueue[TimingWheelSlot],
 	clock hrtime.Clock,
 ) TimingWheel {
 	tw := &timingWheel{
-		ctx:                  ctx,
-		tickMs:               tickMs,
-		startMs:              startMs,
-		slotSize:             slotSize,
-		globalSlotCounterRef: slotCounter,
-		interval:             tickMs * slotSize,
-		currentTimeMs:        startMs - (startMs % tickMs), // truncate the remainder as startMs left boundary
-		slots:                make([]TimingWheelSlot, slotSize),
-		globalDqRef:          dq,
-		clock:                clock,
+		ctx:           ctx,
+		tickMs:        tickMs,
+		startMs:       startMs,
+		slotSize:      slotSize,
+		globalStats:   stats,
+		interval:      tickMs * slotSize,
+		currentTimeMs: startMs - (startMs % tickMs), // truncate the remainder as startMs left boundary
+		slots:         make([]TimingWheelSlot, slotSize),
+		globalDqRef:   dq,
+		clock:         clock,
 	}
-	// Slot initialize by doubly linked list.
+	// Slot initializes by doubly linked list.
 	for i := int64(0); i < slotSize; i++ {
 		tw.slots[i] = NewXSlot()
 	}
-	tw.globalSlotCounterRef.Add(slotSize)
+	tw.globalStats.RecordSlotCount(slotSize)
 	tw.setOverflowTimingWheel(nil)
 	return tw
 }
@@ -190,8 +189,6 @@ type xTimingWheels struct {
 	twEventPool    *timingWheelEventsPool
 	gPool          *ants.Pool
 	stats          *timingWheelStats
-	taskCounter    *atomic.Int64
-	slotCounter    *atomic.Int64
 	isRunning      *atomic.Bool
 	name           string
 	clock          hrtime.Clock
@@ -204,14 +201,6 @@ func (xtw *xTimingWheels) GetTickMs() int64 {
 
 func (xtw *xTimingWheels) GetStartMs() int64 {
 	return xtw.tw.GetStartMs()
-}
-
-func (xtw *xTimingWheels) GetTaskCounter() int64 {
-	return xtw.taskCounter.Load()
-}
-
-func (xtw *xTimingWheels) GetSlotSize() int64 {
-	return xtw.slotCounter.Load()
 }
 
 func (xtw *xTimingWheels) Shutdown() {
@@ -359,16 +348,14 @@ func (xtw *xTimingWheels) schedule(ctx context.Context) {
 						xtw.handleTask(task)
 					}
 					if op == addTask {
-						xtw.taskCounter.Add(1)
+						xtw.stats.RecordJobAliveCount(1)
 					}
 				case cancelTask:
 					jobID, ok := event.GetCancelTaskJobID()
 					if !ok || cancelDisabled.(bool) {
 						goto recycle
 					}
-					if err := xtw.cancelTask(jobID); err == nil {
-						xtw.taskCounter.Add(-1)
-					}
+					_ = xtw.cancelTask(jobID)
 				case unknown:
 					fallthrough
 				default:
@@ -406,6 +393,7 @@ func (xtw *xTimingWheels) schedule(ctx context.Context) {
 					xtw.advanceClock(slot.GetExpirationMs())
 					// Here related to slot level upgrade and downgrade.
 					slot.Flush(xtw.handleTask)
+					xtw.stats.UpdateSlotActiveCount(xtw.dq.Len())
 				}
 			}
 		}(ctx.Value(disableTimingWheelsScheduleExpiredSlot))
@@ -553,6 +541,7 @@ func (xtw *xTimingWheels) cancelTask(jobID JobID) error {
 
 	defer func() {
 		xtw.stats.IncreaseJobCancelledCount()
+		xtw.stats.RecordJobAliveCount(-1)
 	}()
 
 	task.Cancel()

--- a/timer/x_timing_wheels.go
+++ b/timer/x_timing_wheels.go
@@ -102,7 +102,7 @@ func (tw *timingWheel) addTask(task Task, level int64) error {
 
 	if diff < tickMs {
 		task.setSlot(immediateExpiredSlot)
-		tw.advanceClock(currentTimeMs)
+		//tw.advanceClock(currentTimeMs)
 		return fmt.Errorf("[timing wheel] task task expired ms  %d is before %d, %w",
 			taskExpiredMs, currentTimeMs+tickMs, ErrTimingWheelTaskIsExpired)
 	} else if diff >= tickMs && diff < interval {

--- a/timer/x_timing_wheels_others.go
+++ b/timer/x_timing_wheels_others.go
@@ -31,8 +31,6 @@ func NewTimingWheels(ctx context.Context, opts ...TimingWheelsOption) TimingWhee
 
 	xtw := &xTimingWheels{
 		ctx:          ctx,
-		taskCounter:  &atomic.Int64{},
-		slotCounter:  &atomic.Int64{},
 		stopC:        make(chan struct{}),
 		twEventC:     ipc.NewSafeClosableChannel[*timingWheelEvent](1024),
 		expiredSlotC: ipc.NewSafeClosableChannel[TimingWheelSlot](128),
@@ -60,13 +58,16 @@ func NewTimingWheels(ctx context.Context, opts ...TimingWheelsOption) TimingWhee
 	if tw.slotSize <= 0 {
 		tw.slotSize = 20
 	}
+	if xtw.isStatsEnabled {
+		xtw.stats = newTimingWheelStats(xtw)
+	}
 	xtw.dq = queue.NewArrayDelayQueue[TimingWheelSlot](ctx, 128)
 	xtw.tw = newTimingWheel(
 		ctx,
 		tw.tickMs,
 		tw.slotSize,
 		tw.startMs,
-		xtw.slotCounter,
+		xtw.stats,
 		xtw.dq,
 		xtw.clock,
 	)
@@ -78,9 +79,6 @@ func NewTimingWheels(ctx context.Context, opts ...TimingWheelsOption) TimingWhee
 	if xtw.name == "" {
 		// FIXME UUID
 		xtw.name = "default-" + strconv.FormatInt(xtw.GetStartMs(), 10)
-	}
-	if xtw.isStatsEnabled {
-		xtw.stats = newTimingWheelStats(xtw)
 	}
 	xtw.schedule(ctx)
 	return xtw

--- a/timer/x_timing_wheels_test.go
+++ b/timer/x_timing_wheels_test.go
@@ -38,7 +38,6 @@ func TestTimingWheel_AlignmentAndSize(t *testing.T) {
 	t.Logf("tw tickMs alignment: %d\n", unsafe.Alignof(tw.tickMs))
 	t.Logf("tw startMs alignment: %d\n", unsafe.Alignof(tw.startMs))
 	t.Logf("tw slotSize alignment: %d\n", unsafe.Alignof(tw.slotSize))
-	t.Logf("tw globalSlotCounterRef alignment: %d\n", unsafe.Alignof(tw.globalSlotCounterRef))
 	t.Logf("tw overflowWheelRef alignment: %d\n", unsafe.Alignof(tw.overflowWheelRef))
 	t.Logf("tw globalDqRef alignment: %d\n", unsafe.Alignof(tw.globalDqRef))
 
@@ -48,7 +47,6 @@ func TestTimingWheel_AlignmentAndSize(t *testing.T) {
 	t.Logf("tw tickMs size: %d\n", unsafe.Sizeof(tw.tickMs))
 	t.Logf("tw startMs size: %d\n", unsafe.Sizeof(tw.startMs))
 	t.Logf("tw slotSize size: %d\n", unsafe.Sizeof(tw.slotSize))
-	t.Logf("tw globalSlotCounterRef size: %d\n", unsafe.Sizeof(tw.globalSlotCounterRef))
 	t.Logf("tw overflowWheelRef size: %d\n", unsafe.Sizeof(tw.overflowWheelRef))
 	t.Logf("tw globalDqRef size: %d\n", unsafe.Sizeof(tw.globalDqRef))
 }
@@ -62,8 +60,6 @@ func TestXTimingWheels_AlignmentAndSize(t *testing.T) {
 	t.Logf("tw twEventC alignment: %d\n", unsafe.Alignof(tw.twEventC))
 	t.Logf("tw twEventPool alignment: %d\n", unsafe.Alignof(tw.twEventPool))
 	t.Logf("tw expiredSlotC alignment: %d\n", unsafe.Alignof(tw.expiredSlotC))
-	t.Logf("tw taskCounter alignment: %d\n", unsafe.Alignof(tw.taskCounter))
-	t.Logf("tw slotCounter alignment: %d\n", unsafe.Alignof(tw.slotCounter))
 	t.Logf("tw isRunning alignment: %d\n", unsafe.Alignof(tw.isRunning))
 	t.Logf("tw dq alignment: %d\n", unsafe.Alignof(tw.dq))
 	t.Logf("tw tasksMap alignment: %d\n", unsafe.Alignof(tw.tasksMap))
@@ -75,8 +71,6 @@ func TestXTimingWheels_AlignmentAndSize(t *testing.T) {
 	t.Logf("tw twEventC size: %d\n", unsafe.Sizeof(tw.twEventC))
 	t.Logf("tw twEventPool size: %d\n", unsafe.Sizeof(tw.twEventPool))
 	t.Logf("tw expiredSlotC size: %d\n", unsafe.Sizeof(tw.expiredSlotC))
-	t.Logf("tw taskCounter size: %d\n", unsafe.Sizeof(tw.taskCounter))
-	t.Logf("tw slotCounter size: %d\n", unsafe.Sizeof(tw.slotCounter))
 	t.Logf("tw isRunning size: %d\n", unsafe.Sizeof(tw.isRunning))
 	t.Logf("tw dq size: %d\n", unsafe.Sizeof(tw.dq))
 	t.Logf("tw tasksMap size: %d\n", unsafe.Sizeof(tw.tasksMap))
@@ -92,8 +86,6 @@ func TestNewTimingWheels(t *testing.T) {
 	)
 	t.Logf("tw tickMs: %d\n", tw.GetTickMs())
 	t.Logf("tw startMs: %d\n", tw.GetStartMs())
-	t.Logf("tw slotSize: %d\n", tw.GetSlotSize())
-	t.Logf("tw globalTaskCounterRef: %d\n", tw.GetTaskCounter())
 	<-ctx.Done()
 	time.Sleep(100 * time.Millisecond)
 }
@@ -207,15 +199,11 @@ func TestXTimingWheels_ScheduleFunc_ConcurrentFinite(t *testing.T) {
 
 	t.Logf("tw tickMs: %d\n", tw.GetTickMs())
 	t.Logf("tw startMs: %d\n", tw.GetStartMs())
-	t.Logf("tw slotSize: %d\n", tw.GetSlotSize())
-	t.Logf("tw tasks: %d\n", tw.GetTaskCounter())
 	<-ctx.Done()
 	time.Sleep(100 * time.Millisecond)
-	t.Logf("final tw tasks: %d\n", tw.GetTaskCounter())
 	expectedExecCount := len(delays) * 2
 	actualExecCount := execCounter.Load()
 	assert.Equal(t, expectedExecCount, int(actualExecCount))
-	assert.Equal(t, int64(0), tw.GetTaskCounter())
 }
 
 func TestXTimingWheels_ScheduleFunc_sdkClock_1MsInfinite(t *testing.T) {
@@ -313,8 +301,6 @@ func TestXTimingWheels_AfterFunc_Slots(t *testing.T) {
 	}
 	t.Logf("tw tickMs: %d\n", tw.GetTickMs())
 	t.Logf("tw startMs: %d\n", tw.GetStartMs())
-	t.Logf("tw slotSize: %d\n", tw.GetSlotSize())
-	t.Logf("tw tasks: %d\n", tw.GetTaskCounter())
 
 	<-time.After(100 * time.Millisecond)
 	taskIDs := make([]string, 0, len(delays))

--- a/timer/x_timing_wheels_windows.go
+++ b/timer/x_timing_wheels_windows.go
@@ -30,8 +30,6 @@ func NewTimingWheels(ctx context.Context, opts ...TimingWheelsOption) TimingWhee
 
 	xtw := &xTimingWheels{
 		ctx:          ctx,
-		taskCounter:  &atomic.Int64{},
-		slotCounter:  &atomic.Int64{},
 		stopC:        make(chan struct{}),
 		twEventC:     ipc.NewSafeClosableChannel[*timingWheelEvent](1024),
 		expiredSlotC: ipc.NewSafeClosableChannel[TimingWheelSlot](128),
@@ -59,13 +57,16 @@ func NewTimingWheels(ctx context.Context, opts ...TimingWheelsOption) TimingWhee
 	if tw.slotSize <= 0 {
 		tw.slotSize = 20
 	}
+	if xtw.isStatsEnabled {
+		xtw.stats = newTimingWheelStats(xtw)
+	}
 	xtw.dq = queue.NewArrayDelayQueue[TimingWheelSlot](ctx, 128)
 	xtw.tw = newTimingWheel(
 		ctx,
 		tw.tickMs,
 		tw.slotSize,
 		tw.startMs,
-		xtw.slotCounter,
+		xtw.stats,
 		xtw.dq,
 		xtw.clock,
 	)
@@ -77,9 +78,6 @@ func NewTimingWheels(ctx context.Context, opts ...TimingWheelsOption) TimingWhee
 	if xtw.name == "" {
 		// FIXME UUID
 		xtw.name = "default-" + strconv.FormatInt(xtw.GetStartMs(), 10)
-	}
-	if xtw.isStatsEnabled {
-		xtw.stats = newTimingWheelStats(xtw)
 	}
 	xtw.schedule(ctx)
 	return xtw

